modules/core/src/va_intel.cpp

```diff
@@ -0,0 +1,517 @@
+
+# if (defined(HAVE_VA_INTEL) && defined(HAVE_OPENCL))
+    if (contextInitialized)
+    {
+        cl_mem clBuffer = (cl_mem)u.handle(ACCESS_WRITE);
+
+        using namespace cv::ocl;
+        Context& ctx = Context::getDefault();
+        cl_context context = (cl_context)ctx.ptr();
+
+        cl_int status = 0;
+
+        cl_mem clImageY = clCreateFromVA_APIMediaSurfaceINTEL(context, CL_MEM_READ_ONLY, &surface, 0, &status);
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clCreateFromVA_APIMediaSurfaceINTEL failed (Y plane)");
+        cl_mem clImageUV = clCreateFromVA_APIMediaSurfaceINTEL(context, CL_MEM_READ_ONLY, &surface, 1, &status);
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clCreateFromVA_APIMediaSurfaceINTEL failed (UV plane)");
+
+        cl_command_queue q = (cl_command_queue)Queue::getDefault().ptr();
+
+        cl_mem images[2] = { clImageY, clImageUV };
+        status = clEnqueueAcquireVA_APIMediaSurfacesINTEL(q, 2, images, 0, NULL, NULL);
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clEnqueueAcquireVA_APIMediaSurfacesINTEL failed");
+        if (!ocl::ocl_convert_nv12_to_bgr(clImageY, clImageUV, clBuffer, (int)u.step[0], u.cols, u.rows))
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: ocl_convert_nv12_to_bgr failed");
+        status = clEnqueueReleaseVA_APIMediaSurfacesINTEL(q, 2, images, 0, NULL, NULL);
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clEnqueueReleaseVA_APIMediaSurfacesINTEL failed");
+
+        status = clFinish(q); // TODO Use events
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clFinish failed");
+
+        status = clReleaseMemObject(clImageY); // TODO RAII
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clReleaseMem failed (Y plane)");
+        status = clReleaseMemObject(clImageUV);
+        if (status != CL_SUCCESS)
+            CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: clReleaseMem failed (UV plane)");
+    }
+    else
+# endif // HAVE_VA_INTEL && HAVE_OPENCL
+    {
+        VAStatus status = 0;
+
+        status = vaSyncSurface(display, surface);
+        if (status != VA_STATUS_SUCCESS)
+            CV_Error(cv::Error::StsError, "VA-API: vaSyncSurface failed");
+
+        VAImage image;
+        status = vaDeriveImage(display, surface, &image);
+        if (status != VA_STATUS_SUCCESS)
+            CV_Error(cv::Error::StsError, "VA-API: vaDeriveImage failed");
+
+        unsigned char* buffer = 0;
+        status = vaMapBuffer(display, image.buf, (void **)&buffer);
+        if (status != VA_STATUS_SUCCESS)
+            CV_Error(cv::Error::StsError, "VA-API: vaMapBuffer failed");
+
+        CV_Assert(image.format.fourcc == VA_FOURCC_NV12);
+
+        Mat m = u.getMat(ACCESS_WRITE);
```

Comment:
@apavlenko, @alalek, please clarify - should this change be applied also in case of InputArray src=Mat ?
Also, it seems that the logic of function can be changed instead - from
UMat u = dst.getUMat();
if (contextInitialized) { ... } else { u.release(); Mat m = dst.getMat(ACCESS_WRITE); }
to
if (contextInitialized) { UMat u = dst.getUMat(); } else { Mat m = dst.getMat(ACCESS_WRITE); }
Is it correct?